avalon/nuke/lib.py

```diff
@@ -3,82 +3,283 @@
+        {"name": 'avalon_data',
+         "value": 'Warning! Do not change following data!',
+         "type": "Text_Knob"},
+        {"name": '__divider__'},
+        {"name": 'begin',
+         "value": 'Avalon data group',
+         "type": "Tab_Knob", "group": -1}
+    ]
+    non_hiden = ["asset", "subset", "name", "namespace"]
+
     try:
-        node['avalon'].value()
-    except Exception:
-        tab = nuke.Tab_Knob("Avalon")
-        uk = nuke.Text_Knob('avalon', 'avalon data')
-        node.addKnob(tab)
-        node.addKnob(uk)
-        log.info("created new user knob avalon")
+        # create Avalon Tab and basic knobs
+        for k in knobs[:-1]:
+            if not k.get("group"):
+                if "__divider__" in k["name"]:
+                    knob = nuke.Text_Knob("")
+                    node.addKnob(knob)
+                elif k["name"] not in node.knobs().keys():
+                    knob = eval("nuke.{type}('{name}')".format(**k))
+                    node.addKnob(knob)
+                    try:
+                        knob.setValue(k['value'])
+                    except TypeError as E:
+                        print(E)
+            else:
+                if k["name"] not in node.knobs().keys():
+                    knob = eval((
+                        "nuke.{type}('{name}',"
+                        "'{value}', {group})"
+                    ).format(**k))
+                    node.addKnob(knob)
+
+        # add avalon knobs for imprinting data
+        for k, v in data.items():
+            label = k
+            name = prefix + label
+            value = str(v)
+            if label in non_hiden:
+                if name not in node.knobs().keys():
+                    log.info("Setting: `{0}` to `{1}`".format(name, value))
+                    knob = eval((
+                        "nuke.String_Knob('{name}','{label}',"
+                        "'{value}')"
+                    ).format(
+                        name=name,
+                        label=label,
+                        value=value
+                    ))
+                    node.addKnob(knob)
+                else:
+                    log.info("Updating: `{0}` to `{1}`".format(name, value))
+                    node[name].setValue(value)
+            else:
+                if name not in node.knobs().keys():
+                    log.info("Setting: `{0}` to `{1}`".format(name, value))
+                    knob = eval((
+                        "nuke.Text_Knob('{name}','{label}',"
+                        "'{value}')"
+                    ).format(
+                        name=name,
+                        label=label,
+                        value=value
+                    ))
+                    node.addKnob(knob)
+                else:
+                    log.info("Updating: `{0}` to `{1}`".format(name, value))
+                    node[name].setValue(str(value))
+
+        # adding closing group knob
+        knob = eval(
```

Comment:
Why is this using `eval()`? That seems very fishy that it's required to do it like that.

I'd think that the pythonic way would be:
```python
fn = getattr(nuke, knob["type"])
knob = fn(name, value, group)
```

Actually, even line 132 to 161 could be replaced with:
```python
# pseudocode
knob = node.knob(name)
if not knob:
    # Create knob
    knob_type = "String_Knob" if label in non_hidden else "Text_Knob"
    fn = getattr(nuke, knob_type)
    knob = fn(name, label, value)
knob.setValue(value)
```